HIPs/firefly.md

```diff
@@ -0,0 +1,374 @@
+
+The pluggability of the architecture allows other layers in the stack to have different scalability goals or constraints. For instance, if PostgreSQL is chosen as the database for a deployment of FireFly Core, PostgreSQL might be configured for replicated HA (via Patroni etc.). Then the blockchain event streaming service for your chosen blockchain technology might have a different HA strategy, requiring active/passive failover for example. This inherent pluggability is core to the architecture of FireFly.
+
+#### Leader Election
+
+One of the critical tasks of FireFly is to sequence events. To order events that are sent by applications through FireFly into the network, and most importantly to deliver those events back to applications in the exact order that has been agreed via pinning to a blockchain. These sequencing requirements mean there are some processing jobs that must run as a singleton instance within the FireFly core cluster. Examples include batch assembly to pack 100s of messages being sent off-chain into a single pinning transaction on the blockchain, and delivery of events to an individual application instance when the application itself is horizontally scaled. For these tasks, FireFly requires leader election when running in n-way cluster mode. Leader election is itself a pluggable interface.
+
+### End User Features
+
+#### Developer friendly REST+WebSocket API
+
+The FireFly API provides building blocks designed to solve 90% of the needs of a multi-party system, without expensive and error prone custom development of on-chain smart contracts. When custom on-chain smart contract logic is required, the API allows it to be invoked via the same event-driven programming paradigm at the application level.
+
+Developers build to this convenient REST API using their existing API/web/mobile development skill-set. They scale their application using normal n-way cluster scale approaches, and can think about the multi-party system tier in the same way they think about their other infrastructure components, like databases and messaging systems.
+
+The model is event-driven, encouraging application design that allows robust business transactions to execute within a multi-party system. By embracing this event-driven model through simple HTTP and WebSockets in the application tier, responsive user experiences come naturally.
+
+#### Building blocks for enterprise multi-party systems
+
+Enterprise multi-party systems have a number of common attributes, which are distinct from the patterns prevalent in consumer blockchain applications.
+
+Each of the following is encompassed in the design of FireFly:
+
+- Data privacy by default
+  - Regulatory requirements on privacy of data
+  - Regulatory requirements over data retention and deletion
+  - Competitive value of business data
+  - Competitive value of transaction pattern analysis, and other metadata leakage
+  - Limits the use of complex on-chain business logic
```

Comment:
How is this an intrinsic property of FireFly? It depends on the application, no?

Reply:
this is referring to the fact that FireFly itself is dependent on on-chain logic for the foundational functions such as hash pinning and event publishing. A typical solution built on FireFly therefore does not need to have custom on-chain logic any longer.

Reply:
If that is the case, the blockchain is used just as a time stamping service. What's the point of the blockchain? Hashes can be published on The New York Times.

Reply:
time-stamping via hash pinning is utilized more for ordering purposes rather than data immutability, see my response to your later question here https://github.com/hyperledger/hyperledger-hip/pull/3#discussion_r644772585 for more details